### DISCUSSION TITLE
7490 / 07.28 / openssh / Unable to connect

### DISCUSSION THREAD

**Discussion by rozwad** [Freetz-NG/freetz-ng#379] [Allgemein]
Recently I have compiled Freetz-NG image for FB 7490 based on AVM firmware 7.28 rev90133 (all regions) including OpenSSH package, version 8.7p1. Have upgraded FB from legacy Freetz based on 7.12 firmware with OpenSSH 8.1p1. For NG I replicated the config as much as possible from my previous Freetz config.

Unfortunately I am unable to connect to the upgraded OpenSSH 8.7p1 server – the server is simply "silent" i.e. does not respond, the clients I use to connect to 7490 do no show any messages. Had to recompile another Freetz-NG image with Dropbear to login to FB with SSH. Checked in terminal – `netstat -tlnp` shows that FB listens to the port number configured for OpenSSH. There are no logs related to OpenSSH daemon (sshd) in `/var/log/messages`. 

Does anyone else has experienced similar issue with OpenSSH in Freetz-NG or elsewhere?
Would appreciate any hints how to troubleshoot that.

Best regards.

**Comment by fda77** [comment]
You could check syslog by webif.
Maybe you client is to old? Freetz/ was not updated for some years and newer versions of tools with crypto methods disabled unsecure encryptions.
To see what the daemon really does you could add strace package and: `strace -fff -p PID-OF-SSHD` (or similar)

**Reply by rozwad** [reply]
Thank you, @fda77, for your response.
Syslog does not contain any logs for OpenSSH/SSHD, neither `strace` shows any messages at connection attempts to OpenSSH server, just attachment confirmation to SSHD process,

- at the first `strace` launch:
```
strace: Process 4819 attached
futex(0x65ca9c, FUTEX_WAIT_PRIVATE, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=4991, si_uid=0, si_status=1, si_utime=35, si_stime=2} ---
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 1}], WNOHANG, NULL) = 4991
wait4(-1, 0x7fcdb5e0, WNOHANG, NULL)    = -1 ECHILD (No child processes)
sigreturn({mask=[]})                    = 4238
futex(0x65ca9c, FUTEX_WAIT_PRIVATE, 0, NULL
```
- at the next `strace` runs (looks like a truncated output of the above one):
```
strace: Process 4819 attached
futex(0x65ca9c, FUTEX_WAIT_PRIVATE, 0, NULL
```
- after OpenSSH restart:
```
strace: Process 6027 attached
pselect6(5, [3 4], NULL, NULL, NULL, {[], 16}
```

I am using the latest PuTTY version 0.76.

Would appreciate any further hints on the troubleshooting.

**Reply by fda77** [reply]
This strace is very short. Can you connect locally? I have no really an idea in which part it does not work

**Reply by rozwad** [reply]
Same as for the remote connection attempts – I am unable to connect locally to the OpenSSH server – simply there is no response to `ssh locahost -p 222` and `strace -fff -p PID-OF-SSH` shows only the process attachment confirmation, Just dropbear server works fine (checked for the same ports).

`ps | grep sshd` on the Freetz-NG image with the non-responding OpenSSH 8.7p1 shows `sshd: /usr/sbin/sshd -p 222 -f /mod/etc/openssh.conf [listener] 0 of 10-100`, while on the legacy Freetz image with a working OpenSSH 8.1p1 shows just `sshd: /usr/sbin/sshd -p 222 -f /mod/etc/openssh.conf` (i.e. without `[listener] 0 of 10-100`).
Is that as expected?

Please also see the attached debug logs for the SSH server and client connection attempt:

- [ssh-server.debug.log](https://github.com/Freetz-NG/freetz-ng/files/7172183/ssh-server.debug.log)
- [ssh-client.debug.log](https://github.com/Freetz-NG/freetz-ng/files/7172184/ssh-client.debug.log)

For some reason the SSH server gets stuck at `preauth child monitor started` and does not respond to `SSH2_MSG_KEXINIT`. Any idea how to troubleshoot that further?

**Reply by fda77** [reply]
Freetz/ has much older versions of libs and the progam itself, so there could be changes. "SSH2_MSG_KEXINIT" seems a good startpoint.
Googles 1st hit: https://stackoverflow.com/questions/2419412/ssh-connection-stop-at-debug1-ssh2-msg-kexinit-sent
Maybe your hostkey (of the server/box) is to old?
Here is another which stuck after an update with "SSH2_MSG_KEXINIT": https://superuser.com/questions/1657052/after-debian-8-to-debian-9-to-debian-10-dist-upgrade-i-cannot-connect-via-ssh-as
Or from https://bugs.archlinux.org/task/59838
```
> From the release notes this seems the most likely noted change
* sshd(8): the semantics of PubkeyAcceptedKeyTypes and the similar
HostbasedAcceptedKeyTypes options have changed. These now specify
signature algorithms that are accepted for their respective
authentication mechanism, where previously they specified accepted
key types. This distinction matters when using the RSA/SHA2
signature algorithms "rsa-sha2-256", "rsa-sha2-512" and their
certificate counterparts. Configurations that override these
options but omit these algorithm names may cause unexpected
authentication failures (no action is required for configurations
that accept the default for these options).
```

**Reply by rozwad** [reply]
Summary of some further observations:
- Regenerated all the host keys → **NOK/cannot** connect.
- Reverted commits for the OpenSSH version bumps, recompiled for OpenSSH 8.6p1 and 8.5p1 → **NOK/cannot** connect.
- Copied `sshd` binary with OpenSSH 8.7p1 from Freetz-NG into a legacy Freetz image based on 07.1X (checked on 7490 and 7590), and started another server from shell → **NOK/cannot** connect.
- Copied `sshd` binary with OpenSSH 8.1p1 from legacy Freetz image into the Freetz-NG based on 07.28, and started another server from shell → **OK/can** connect.

Will try to revert further _bump openssh_ commits and recompile Freetz-NG to check what is the latest OpenSSH version which works.

**Reply by fda77** [reply]
To use an older version it should be enough to change make/PKG/PKG.mk and replace make/PKG/patches/ dir.
Maybe its related to openssl, with ng you could switch between 3 versions, freetz/ has only (no longer supported by upstream) 1.0

**Reply by rozwad** [reply]
Thank you for the tip. Version 1.0.2u (deprecated) works fine. Found a [KB article](https://knowledgebase.progress.com/articles/Article/Support-for-OpenSSL-1-1-1) which tells, that there are several "ciphers that are reported in the build of OpenSSL 1.0.2 that are not reported in the build of OpenSSL 1.1.1, so these can be considered unsupported by OpenSSL 1.1.1." Maybe that is the root cause? Or is it rather something related to OpenSSH server default config (or misconfiguration)?

Tried also to compile OpenSSL 3.0, however I'm getting OpenSSH compilation errors for multiple files (`OSSL_DEPRECATEDIN_3_0` for `EC_KEY` functions or expressions within OpenSSH code). Looks like a wider OpenSSL 3.0 and OpenSSH 8.x incompatibility issue.

**Reply by fda77** [reply]
See my 1st answer here, Openssl 1.1 has some unsafe/insecure things disabled which here okay with version 1.0. There are chipher, hash methods etc. And the length of the keys could be cause such things (if i remember correctly, something with 1024 is no more okay and i had to genereate new with 2048)
If you have additional the openssl binary installed, it can tell you which algorithms are available
